### PR TITLE
`librdkafka` as submodule

### PR DESCRIPTION
**Pull request by felixschlegel**
### Motivation

Previously users of our Package had to install `librdkafka` through their package managers to be able to run this package, as we relied on linking `librdkafka` **dynamically**.

With this PR, we ship `librdkafka` as a **git submodule** that is then built with SPM. This overcomes the issues we previously had with dynamic linking (e.g. not finding header files etc.).

### Modifications:

* add `librdkafka` submodule to project
* create a hardcoded `config.h` files that contain information on how
  `librdkafka` should be compiled (normally this is done through a the
`librdkafka/configure.sh` script, though we cannot invoke this at
buildtime with SPM)
* add package dependencies to `zstd` and `swift-nio-ssl`
* build `librdkafka` package with SPM instead of relying on a
  `.systemLibary` target
* create bridge between `<openssl/*>` headers and `swift-nio-ssl` with
  `custom` include folder
* update Docker to install `zlib`
* remove librdkafka installation from Docker
* create hard-coded `config.h` files used to compile librdkafka
* automatically update submodule from Docker

### Known Issues

* our Package assumes that `zlib`, `curl` and `sasl` are already installed on the system we are building on
* the new build process has **not** been tested on **x86_64**
* to build `librdkafka`, a config script has to be run that creates a file called `config.h` that configures the `lirbdkafka` build according to the underlying system, as we cannot run a build-script with SPM (I think 😅) I hardcoded a couple of `config.h` files to build `librdkafka` for the most used arch/os combinations (See: [Building librdkafka](https://github.com/confluentinc/librdkafka#building))
* at the moment, our [facebook/zstd](https://github.com/facebook/zstd) dependency throws out a lot of build warnings, however, this is a know issue: https://github.com/facebook/zstd/issues/3328

### CODE REVIEW

**Review by FranzBusch — CHANGES_REQUESTED** [review]
This is great! It will make working with this package a lot easier. I left some inline comments that we need to address. Furthermore, we need to add a notice file similar to [this](https://github.com/apple/swift-cassandra-client/blob/main/NOTICE.txt)

File: Sources/Crdkafka/librdkafka
[the remark this answers is not in the document]
Comment (felixschlegel) [reply Sources/Crdkafka/librdkafka:1]: Yes, now we are pointing at [librdkafka-v2.1.1](https://github.com/confluentinc/librdkafka/releases/tag/v2.1.1)

File: Package.swift [hunk 1]
```diff
@@ -32,8 +61,33 @@ let package = Package(
     dependencies: [
         .package(url: "https://github.com/apple/swift-nio.git", from: "2.43.1"),
         .package(url: "https://github.com/apple/swift-log.git", from: "1.0.0"),
+        .package(url: "https://github.com/apple/swift-nio-ssl.git", .upToNextMajor(from: "2.21.0")),
```
[the remark this answers is not in the document]
Comment (felixschlegel) [reply Package.swift]: I talked to @lukasa, and our new approach is to have `openssl` as a `.systemLibrary` target.

File: Package.swift [hunk 2]
```diff
@@ -32,8 +61,33 @@ let package = Package(
...
+        // The zstd Swift package produces warnings that we cannot resolve:
```
[the remark this answers is not in the document]
Comment (felixschlegel) [reply Package.swift:55]: Just warnings ⚠️ :

```bash
/swift-kafka-gsoc/Sources/Crdkafka/librdkafka/src/rdkafka_zstd.c:38:2: warning: definition of configuration macro 'ZSTDLIB_VISIBLE' has no effect on the import of 'libzstd.errors'; pass '-DZSTDLIB_VISIBLE=...' on the command line to configure the module [-Wconfig-macros]
#include <zstd_errors.h>
 ^
/swift-kafka-gsoc/.build/checkouts/zstd/lib/zstd.h:28:13: note: macro was defined here
#    define ZSTDLIB_VISIBLE __attribute__ ((visibility ("default")))
            ^
/swift-kafka-gsoc/Sources/Crdkafka/librdkafka/src/rdkafka_zstd.c:38:2: warning: definition of configuration macro 'ZSTDLIB_HIDDEN' has no effect on the import of 'libzstd.errors'; pass '-DZSTDLIB_HIDDEN=...' on the command line to configure the module [-Wconfig-macros]
#include <zstd_errors.h>
 ^
/swift-kafka-gsoc/.build/checkouts/zstd/lib/zstd.h:36:13: note: macro was defined here
#    define ZSTDLIB_HIDDEN __attribute__ ((visibility ("hidden")))
            ^
/swift-kafka-gsoc/Sources/Crdkafka/librdkafka/src/rdkafka_zstd.c:38:2: warning: definition of configuration macro 'ZSTD_CLEVEL_DEFAULT' has no effect on the import of 'libzstd.errors'; pass '-DZSTD_CLEVEL_DEFAULT=...' on the command line to configure the module [-Wconfig-macros]
#include <zstd_errors.h>
 ^
/swift-kafka-gsoc/.build/checkouts/zstd/lib/zstd.h:129:11: note: macro was defined here
#  define ZSTD_CLEVEL_DEFAULT 3
          ^
/swift-kafka-gsoc/Sources/Crdkafka/librdkafka/src/rdkafka_zstd.c:38:2: warning: definition of configuration macro 'ZDICTLIB_VISIBLE' has no effect on the import of 'libzstd.errors'; pass '-DZDICTLIB_VISIBLE=...' on the command line to configure the module [-Wconfig-macros]
#include <zstd_errors.h>
 ^
/swift-kafka-gsoc/.build/checkouts/zstd/lib/zdict.h:28:13: note: macro was defined here
#    define ZDICTLIB_VISIBLE __attribute__ ((visibility ("default")))
            ^
/swift-kafka-gsoc/Sources/Crdkafka/librdkafka/src/rdkafka_zstd.c:38:2: warning: definition of configuration macro 'ZDICTLIB_HIDDEN' has no effect on the import of 'libzstd.errors'; pass '-DZDICTLIB_HIDDEN=...' on the command line to configure the module [-Wconfig-macros]
#include <zstd_errors.h>
 ^
/swift-kafka-gsoc/.build/checkouts/zstd/lib/zdict.h:36:13: note: macro was defined here
#    define ZDICTLIB_HIDDEN __attribute__ ((visibility ("hidden")))
            ^
/swift-kafka-gsoc/Sources/Crdkafka/librdkafka/src/rdkafka_zstd.c:38:2: warning: definition of configuration macro 'ZSTDERRORLIB_VISIBLE' has no effect on the import of 'libzstd.errors'; pass '-DZSTDERRORLIB_VISIBLE=...' on the command line to configure the module [-Wconfig-macros]
#include <zstd_errors.h>
 ^
/swift-kafka-gsoc/.build/checkouts/zstd/lib/zstd_errors.h:28:13: note: macro was defined here
#    define ZSTDERRORLIB_VISIBLE __attribute__ ((visibility ("default")))
            ^
/swift-kafka-gsoc/Sources/Crdkafka/librdkafka/src/rdkafka_zstd.c:38:2: warning: definition of configuration macro 'ZSTDERRORLIB_HIDDEN' has no effect on the import of 'libzstd.errors'; pass '-DZSTDERRORLIB_HIDDEN=...' on the command line to configure the module [-Wconfig-macros]
#include <zstd_errors.h>
 ^
/swift-kafka-gsoc/.build/checkouts/zstd/lib/zstd_errors.h:36:13: note: macro was defined here
#    define ZSTDERRORLIB_HIDDEN __attribute__ ((visibility ("hidden")))
```

File: Package.swift [hunk 3]
```diff
@@ -15,6 +15,24 @@
...
+
+let operatingSystem: String
+#if os(macOS)
+operatingSystem = "macOS"
+#elseif os(Linux)
+operatingSystem = "linux"
+#else
+fatalError("This package only supports macOS and Linux at the moment")
+#endif
+
```
[the remark this answers is not in the document]
Comment (felixschlegel) [reply Package.swift]: The `.when(...)` approach works for the target platform, though I have not yet found a better way to find out the target `arch` in SwiftPM 😅

**Comment by felixschlegel** [comment]
cc: @lukasa @gjcairo

I have implemented the changes that @FranzBusch has requested. Regarding the [`NOTICE.txt`](NOTICE.txt) file, I don't know if that is legally correct so I would be happy if you could double-check that 😅 

Cheers,
Felix

File: Package.swift [hunk 4]
```diff
@@ -32,8 +52,32 @@ let package = Package(
...
+            cSettings: [
+                // dummy folder, because config.h is included as "../config.h" in librdkafka
+                .headerSearchPath("./custom/config/\(arch)/macOS/dummy", .when(platforms: [.macOS])),
+                .headerSearchPath("./custom/config/\(arch)/linux/dummy", .when(platforms: [.linux])),
+                .headerSearchPath("./librdkafka/src"),
+            ],
+            linkerSettings: [
+                .linkedLibrary("curl"),
+                .linkedLibrary("sasl2"),
+                .linkedLibrary("z"), // zlib
```
Comment (tomerd) [inline Package.swift:78]: where do these come from? system dependencies?

File: Sources/Crdkafka/custom/config/x86_64/macOS/config.h [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+#define HAVE_STRERROR_R 1
+// strcasestr
+#define HAVE_STRCASESTR 1
+// pthread_setname_darwin
+#define HAVE_PTHREAD_SETNAME_DARWIN 1
+// getrusage
+#define HAVE_GETRUSAGE 1
+// BUILT_WITH
+#define BUILT_WITH "STATIC_LINKING GCC GXX PKGCONFIG OSXLD LIBDL PLUGINS ZLIB SSL SASL_CYRUS ZSTD CURL HDRHISTOGRAM SYSLOG SNAPPY SOCKEM SASL_SCRAM SASL_OAUTHBEARER OAUTHBEARER_OIDC CRC32C_HW"
+#endif /* _CONFIG_H_ */
```
Comment (tomerd) [inline Sources/Crdkafka/custom/config/x86_64/macOS/config.h:110]: are these all generated by kayaks build process ie by running `./configure` for the platform in question?

File: Package.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (tomerd) [reply Package.swift:55]: cc @yim-lee

**Comment by yim-lee** [comment]
@swift-server-bot test this please

**Review by FranzBusch — APPROVED** [review]
LGTM!

**Pull request merged**